### PR TITLE
Automatically convert convolution weights depending on backend

### PR DESCRIPTION
**Pull request by matthiasplappert**
As discussed in https://github.com/fchollet/keras/issues/3307, I recently had trouble with loading the weights of a network that was trained using Theano but evaluated using TensorFlow. While @EderSantana pointed out that this difference in behavior is properly documented in the Wiki, I think we can do better.

My approach includes the backend when storing the weights. This allows us to automatically detect a difference in backend when loading the weights and to convert the kernel weights appropriately. My solution is fully backwards-compatible.

However, there are a couple of things that I'm not sure I got right:
- Should we include `SeparableConvolution2D` in the list of convertible classes?
- Since we include `Convolution1D` and so on, we also convert `Deconvolution1D` and friends since they inherit from the respective `Convolution` layer. Again, I'm not entirely sure if the conversion process is correct in this case.
- Do we need tests for this?

### CODE REVIEW

File: keras/engine/topology.py [hunk 1]
```diff
@@ -2485,6 +2487,20 @@ def load_weights_from_hdf5_group(self, f):
                 weight_value_tuples += zip(symbolic_weights, weight_values)
             K.batch_set_value(weight_value_tuples)
 
+        backend = f.attrs.get('keras_backend')
+        if backend and backend != K._BACKEND:
+            # Weights were originally serialized using a different backend. TensorFlow and Theano
+            # use different implementations for convolutions, which need to be converted.
+            weight_value_tuples = []
+            for layer in flattened_layers:
+                from keras.layers import Convolution1D, Convolution2D, Convolution3D
```
Comment (fchollet) [inline keras/engine/topology.py]: An import in a loop?

File: keras/engine/topology.py [hunk 2]
```diff
@@ -2485,6 +2487,20 @@ def load_weights_from_hdf5_group(self, f):
                 weight_value_tuples += zip(symbolic_weights, weight_values)
             K.batch_set_value(weight_value_tuples)
 
+        backend = f.attrs.get('keras_backend')
+        if backend and backend != K._BACKEND:
+            # Weights were originally serialized using a different backend. TensorFlow and Theano
+            # use different implementations for convolutions, which need to be converted.
+            weight_value_tuples = []
+            for layer in flattened_layers:
```
Comment (fchollet) [inline keras/engine/topology.py]: This loop should be integrated with the first one. Otherwise we need two calls to `batch_set_values`, which is less efficient. Weights should be converted in Numpy-space before being set.

**Comment by matthiasplappert** [comment]
Thanks for your feedback, @fchollet. I've thought about the problem some more and came up with a slightly different implementation in https://github.com/fchollet/keras/pull/3345/commits/b715ecc744b115d3328b0d5a68c1d2e0b8a2c9de that feels less hard-coded to this specific problem. To do so, I introduced a very simple `convert_weights` method in `Layer` that takes a list of weights and the backend that the weights originated from and is expected to return a potentially updated list. This allows other layers that might have similar needs in the future to implement this behavior rather easily. It also decouples `topology.py` from the specific layer implementations.

I was initially unsure if this should be a static method instead, but decided against. However, your input on this is very much appreciated.

This also currently misses proper documentation and tests. Before I go ahead and add those, I just wanted to make sure that this approach seems fine with you, since it, if only very slightly, messes with the core `Layer` class.

File: keras/engine/topology.py [hunk 3]
```diff
@@ -2465,11 +2469,14 @@ def load_weights_from_hdf5_group(self, f):
...
             # which provides a speedup in TensorFlow.
             weight_value_tuples = []
+            backend = f.attrs.get('keras_backend')
             for k, name in enumerate(layer_names):
                 g = f[name]
                 weight_names = [n.decode('utf8') for n in g.attrs['weight_names']]
                 weight_values = [g[weight_name] for weight_name in weight_names]
                 layer = flattened_layers[k]
+                if backend and backend != K._BACKEND:
+                    weight_values = layer.convert_weights(weight_values, backend)
```
Comment (fchollet) [inline keras/engine/topology.py]: It is be preferable to confine changes to this loop instead of introducing a new layer method. Here, we just need a loop that builds the `weight_values` list and flips convolution kernels if `layer` is an instance of a convolution layer.

**Comment by fchollet** [comment]
Closing due to no activity for a while. Overall I'm not convinced this is the way to go, either.

**Comment by jonilaserson** [comment]
One comment regarding converting from Theano to TF in CNNs: usually this conversion also comes with a change to image_dim_ordering from 'th' to 'tf'. Staying in 'th' while doing convolutions in TF would result in a x3 slowness in myh experience. Hence it would be useful if the conversion code would also take that into account when loading weights from a file.  

Another note regarding converting the dim_ordering from 'th' to 'tf' is to remember to change the BatchNormalization axis from 1 to -1.
